fix(accordion): correct trigger semantics and guard keyboard events i…

goetzrobin opened the pull request:
…n editable fields

- Remove incorrect heading semantics from trigger
- Use proper button semantics with aria-expanded/aria-controls
- Validate ARIA structure: trigger must be a button and be wrapped by a semantic heading
- Add robust keyboard guarding: ignore navigation/toggle handling inside inputs,etc.
- Allow modifier combos (Ctrl/Meta/Alt) to pass through keyboard handling
- Prevent default only for relevant keys when the accordion has focus and based on orientation
- textareas accept newlines without toggling
- Expands e2e test coversage:
    - Add ARIA semantics checks (trigger/button, heading wrapper, region with aria-labelledby)
    - Cover single/multiple behavior, horizontal/vertical navigation, Home/End, focus wrapping
    - Verify form input interactions (input/textarea/select/contenteditable) don’t toggle accordion
    - Add inert behavior checks for focusables in closed panels
    - Run axe checks for default stories

Fixes: #792,#833

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our
      guidelines: https://github.com/spartan-ng/spartan/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

### Primitives

- [x] accordion
- [ ] alert
- [ ] alert-dialog
- [ ] aspect-ratio
- [ ] avatar
- [ ] badge
- [ ] breadcrumb
- [ ] button
- [ ] calendar
- [ ] card
- [ ] carousel
- [ ] checkbox
- [ ] collapsible
- [ ] combobox
- [ ] command
- [ ] context-menu
- [ ] data-table
- [ ] date-picker
- [ ] dialog
- [ ] dropdown-menu
- [ ] form-field
- [ ] hover-card
- [ ] icon
- [ ] input
- [ ] input-otp
- [ ] label
- [ ] menubar
- [ ] navigation-menu
- [ ] pagination
- [ ] popover
- [ ] progress
- [ ] radio-group
- [ ] scroll-area
- [ ] select
- [ ] separator
- [ ] sheet
- [ ] skeleton
- [ ] slider
- [ ] sonner
- [ ] spinner
- [ ] switch
- [ ] table
- [ ] tabs
- [ ] textarea
- [ ] toggle
- [ ] toggle-group
- [ ] tooltip
- [ ] typography

### Others

- [ ] trpc
- [ ] nx
- [ ] repo
- [ ] cli

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

See commit message above

Closes #792,#833

## What is the new behavior?

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
We remove the hard coded h3 in favor of prompting the user to wrap the trigger with an h[ANYLEVEL] element themselves. Updates the docs to reflect that

## Other information

Commit:
fix(accordion): correct trigger semantics and guard keyboard events in editable fields

- Remove incorrect heading semantics from trigger
- Use proper button semantics with aria-expanded/aria-controls
- Validate ARIA structure: trigger must be a button and be wrapped by a semantic heading
- Add robust keyboard guarding: ignore navigation/toggle handling inside inputs,etc.
- Allow modifier combos (Ctrl/Meta/Alt) to pass through keyboard handling
- Prevent default only for relevant keys when the accordion has focus and based on orientation
- textareas accept newlines without toggling
- Expands e2e test coversage:
    - Add ARIA semantics checks (trigger/button, heading wrapper, region with aria-labelledby)
    - Cover single/multiple behavior, horizontal/vertical navigation, Home/End, focus wrapping
    - Verify form input interactions (input/textarea/select/contenteditable) don’t toggle accordion
    - Add inert behavior checks for focusables in closed panels
    - Run axe checks for default stories

Fixes: #792,#833

MerlinMoos commented:
Hey @goetzrobin, short feedback on this. I'm not sure if we should but a button into a heading. Headings can only contain phrasing content, and a <button> is interactive content that makes the HTML invalid and harms accessibility.

goetzrobin commented:
> Hey @goetzrobin, short feedback on this. I'm not sure if we should but a button into a heading. Headings can only contain phrasing content, and a is interactive content that makes the HTML invalid and harms accessibility.

Hey @MerlinMoos! I actually had the same gut reaction, but this is from the waria specs:

Each accordion header button is wrapped in an element with role [heading](https://w3c.github.io/aria/#heading) that has a value set for [aria-level](https://w3c.github.io/aria/#aria-level) that is appropriate for the information architecture of the page.
If the native host language has an element with an implicit heading and aria-level, such as an HTML heading tag, a native host language element may be used.
The button element is the only element inside the heading element. That is, if there are other visually persistent elements, they are not included inside the heading element.

Here is the full article for reference: https://www.w3.org/WAI/ARIA/apg/patterns/accordion/

But I guess we could make the role an input so people could turn it off?

Commit:
feat(cli): add accordion trigger migration generator and healthcheck

Commit:
feat(cli): fix linting issues and add accordion trigger migration to health check

ashley-hunter reviewed (approved):
Looks good to me! Thank you! We have previously also usually added the healthcheck generator to the generators.json file so someone could run it standalone as a generator, so I guess we could add it in there too, but apart from that all looks great to me! Thanks!

Commit:
feat: reorder generators/schematics to be alphabetic, add missing ones

reordered schematics/generators to be alphabetic.
added missing schematics/generators, e.g. if generators had one and was
missing from schematics and vice versa
added migrate-brain-accordion-trigger
